pkg/operator/controller/controller.go

```diff
@@ -203,6 +193,32 @@ func (r *reconciler) reconcile() (reconcile.Result, error) {
 	return reconcile.Result{}, utilerrors.NewAggregate(errs)
 }
 
+// ensureIngressDeleted tries to delete ingress, and if successful, will remove
+// the finalizer.
+func (r *reconciler) ensureIngressDeleted(ingress *ingressv1alpha1.ClusterIngress, dnsConfig *configv1.DNS) error {
+	// TODO: This should also be tied to the HA type.
+	err := r.finalizeLoadBalancerService(ingress, dnsConfig)
+	if err != nil {
+		return fmt.Errorf("failed to finalize load balancer service for %s: %v", ingress.Name, err)
+	}
+	logrus.Infof("finalized load balancer service for ingress %s", ingress.Name)
+	err = r.ensureRouterDeleted(ingress)
+	if err != nil {
+		return fmt.Errorf("failed to delete deployment for ingress %s: %v", ingress.Name, err)
+	}
+	logrus.Infof("deleted deployment for ingress %s", ingress.Name)
+	// Clean up the finalizer to allow the clusteringress to be deleted.
+	updated := ingress.DeepCopyObject().(*ingressv1alpha1.ClusterIngress)
```

Comment:
Why use `DeepCopyObject` and a cast instead of using `DeepCopy`?

Reply:
Too much time staring at `runtime.Object`... fixed